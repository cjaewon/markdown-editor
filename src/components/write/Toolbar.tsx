import * as React from 'react';
import styled, { css } from 'styled-components';
import {
  MdFormatBold,
  MdFormatItalic,
  MdFormatUnderlined,
  MdFormatStrikethrough,
  MdInsertLink,
  MdFormatQuote,
  MdImage,
  MdCode,
} from 'react-icons/md';
import palette from '../../lib/styles/palette';
import zIndexes from '../../lib/styles/zIndexes';
import { mediaQuery } from '../../lib/styles/media';

// box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.09);
const ToolbarBlock = styled.div<{
  shadow: boolean;
  forMarkdown: boolean;
}>`
  width: 100%;
  top: 0;
  display: flex;
  align-items: center;
  margin-bottom: 1rem;
  width: 100%;
  background: white;
  z-index: ${zIndexes.Toolbar};
  transition: all 0.125s ease-in;
  flex-wrap: wrap;

  ${props =>
    props.forMarkdown &&
    css`
      margin-bottom: 1rem;
      padding-left: 3rem;
      padding-right: 3rem;

      ${mediaQuery(767)} {
        padding-left: 1rem;
        padding-right: 1rem;
        flex-wrap: unset;
        overflow-x: auto;
      }
      width: auto;
    `}

  ${props =>
    props.shadow &&
    css`
      box-shadow: 0px 2px 8px rgba(0, 0, 0, 0.09);
      margin-bottom: 0;
    `}

  .mobile-placeholder {
    display: none;
    ${mediaQuery(767)} {
      display: block;
    }
    height: 1px;
    width: 1rem;
    flex-shrink: 0;
  }
`;

// const ToolbarGroup = styled.div`
//   display: flex;
//   height: 100%;
// `;

const Heading = styled.div`
  font-size: 1rem;
  font-weight: bold;
  font-family: serif;
  span {
    font-size: 0.75rem;
  }
`;
const ToolbarItem = styled.button`
  width: 3rem;
  height: 3rem;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 1.75rem;
  color: ${palette.gray6};
  cursor: pointer;
  flex-shrink: 0;

  ${mediaQuery(767)} {
    width: 2.5rem;
    height: 2.5rem;
    font-size: 1.45rem;
  }
  &:hover {
    color: ${palette.gray9};
    background: ${palette.gray0};
  }
  background: none;
  outline: none;
  border: none;
  padding: 0;

  &.ql-active {
    color: ${palette.teal6};
  }
`;

const Separator = styled.div`
  width: 1px;
  height: 1.25rem;
  margin-left: 0.5rem;
  margin-right: 0.5rem;
  background: ${palette.gray4};
`;
export interface ToolbarProps {
  shadow: boolean;
  mode: 'MARKDOWN' | 'WYSIWYG';
  onClick?: Function;
  onConvert?: () => void;
  innerRef?: React.RefObject<HTMLDivElement>;
  ios?: boolean;
}

const Toolbar: React.FC<ToolbarProps> = ({
  shadow,
  mode,
  onClick = () => {},
  onConvert,
  innerRef,
  ios,
}) => {
  const forMarkdown = mode === 'MARKDOWN';
  return (
    <ToolbarBlock
      id="toolbar"
      shadow={shadow}
      forMarkdown={forMarkdown}
      ref={innerRef}
    >
      <ToolbarItem
        className="ql-header"
        value={1}
        onClick={() => onClick('heading1')}
      >
        <Heading>
          H<span>1</span>
        </Heading>
      </ToolbarItem>
      <ToolbarItem
        className="ql-header"
        value={2}
        onClick={() => onClick('heading2')}
      >
        <Heading>
          H<span>2</span>
        </Heading>
      </ToolbarItem>
      <ToolbarItem
        className="ql-header"
        value={3}
        onClick={() => onClick('heading3')}
      >
        <Heading>
          H<span>3</span>
        </Heading>
      </ToolbarItem>
      <ToolbarItem
        className="ql-header"
        value={4}
        onClick={() => onClick('heading4')}
      >
        <Heading>
          H<span>4</span>
        </Heading>
      </ToolbarItem>
      <Separator />
      {!ios && (
        <>
          <ToolbarItem className="ql-bold" onClick={() => onClick('bold')}>
            <MdFormatBold />
          </ToolbarItem>
          <ToolbarItem className="ql-italic" onClick={() => onClick('italic')}>
            <MdFormatItalic />
          </ToolbarItem>
          {!forMarkdown && (
            <ToolbarItem className="ql-underline">
              <MdFormatUnderlined />
            </ToolbarItem>
          )}
          <ToolbarItem className="ql-strike" onClick={() => onClick('strike')}>
            <MdFormatStrikethrough />
          </ToolbarItem>
          <Separator />
          <ToolbarItem
            className="ql-blockquote"
            onClick={() => onClick('blockquote')}
          >
            <MdFormatQuote />
          </ToolbarItem>
          <ToolbarItem className="ql-link" onClick={() => onClick('link')}>
            <MdInsertLink />
          </ToolbarItem>
        </>
      )}
      <ToolbarItem className="ql-image" onClick={() => onClick('image')}>
        <MdImage />
      </ToolbarItem>

      {!ios && (
        <ToolbarItem
          className="ql-code-block"
          onClick={() => onClick('codeblock')}
        >
          <MdCode />
        </ToolbarItem>
      )}
      <div className="mobile-placeholder"></div>
      {/* <Separator /> */}
      {/* <ToolbarGroup> */}
      {/* forMarkdown ? (
          <ToolbarItem data-testid="quillconvert" onClick={onConvert}>
            <MdRemoveRedEye />
          </ToolbarItem>
        ) : (
          <ToolbarItem data-testid="mdconvert" onClick={onConvert}>
            <FaMarkdown />
          </ToolbarItem>
        ) */}
      {/* </ToolbarGroup> */}
    </ToolbarBlock>
  );
};

export default Toolbar;
