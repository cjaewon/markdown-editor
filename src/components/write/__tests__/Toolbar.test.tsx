import * as React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Toolbar, { ToolbarProps } from '../Toolbar';

describe('Toolbar', () => {
  const setup = (props: Partial<ToolbarProps> = {}) => {
    const initialProps: ToolbarProps = {
      shadow: false,
      mode: 'MARKDOWN',
      onClick: () => {},
    };
    const utils = render(<Toolbar {...initialProps} {...props} />);
    return {
      ...utils,
    };
  };
  it('renders properly', () => {
    setup();
  });
  it('matches snapshot', () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });
  it('onClick works', () => {
    const onClick = jest.fn();
    const utils = setup({
      onClick,
    });
    const headers = utils.container.querySelectorAll('.ql-header');
    expect(headers.length).toBe(4);
    const heading1 = headers[0];
    fireEvent.click(heading1);
    expect(onClick).toBeCalledWith('heading1');
  });
  // it('calls onConvert in markdown mode', () => {
  //   const onConvert = jest.fn();
  //   const { getByTestId } = setup({
  //     onConvert,
  //   });
  //   const convertButton = getByTestId('quillconvert');
  //   fireEvent.click(convertButton);
  //   expect(onConvert).toBeCalled();
  // });
  // it('calls onConvert in markdown mode', () => {
  //   const onConvert = jest.fn();
  //   const { getByTestId } = setup({
  //     mode: 'WYSIWYG',
  //     onConvert,
  //   });
  //   const convertButton = getByTestId('mdconvert');
  //   fireEvent.click(convertButton);
  //   expect(onConvert).toBeCalled();
  // });
});
