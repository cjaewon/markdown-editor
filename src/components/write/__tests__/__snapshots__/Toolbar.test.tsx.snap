// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbar matches snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa eOpsME"
    id="toolbar"
  >
    <button
      class="sc-bxivhb cshiNX ql-header"
      value="1"
    >
      <div
        class="sc-htpNat jrmEoq"
      >
        H
        <span>
          1
        </span>
      </div>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-header"
      value="2"
    >
      <div
        class="sc-htpNat jrmEoq"
      >
        H
        <span>
          2
        </span>
      </div>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-header"
      value="3"
    >
      <div
        class="sc-htpNat jrmEoq"
      >
        H
        <span>
          3
        </span>
      </div>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-header"
      value="4"
    >
      <div
        class="sc-htpNat jrmEoq"
      >
        H
        <span>
          4
        </span>
      </div>
    </button>
    <div
      class="sc-ifAKCX eqJNNy"
    />
    <button
      class="sc-bxivhb cshiNX ql-bold"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.6 10.79c.97-.67 1.65-1.77 1.65-2.79 0-2.26-1.75-4-4-4H7v14h7.04c2.09 0 3.71-1.7 3.71-3.79 0-1.52-.86-2.82-2.15-3.42zM10 6.5h3c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5h-3v-3zm3.5 9H10v-3h3.5c.83 0 1.5.67 1.5 1.5s-.67 1.5-1.5 1.5z"
        />
      </svg>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-italic"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 4v3h2.21l-3.42 8H6v3h8v-3h-2.21l3.42-8H18V4z"
        />
      </svg>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-strike"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 19h4v-3h-4v3zM5 4v3h5v3h4V7h5V4H5zM3 14h18v-2H3v2z"
        />
      </svg>
    </button>
    <div
      class="sc-ifAKCX eqJNNy"
    />
    <button
      class="sc-bxivhb cshiNX ql-blockquote"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 17h3l2-4V7H5v6h3zm8 0h3l2-4V7h-6v6h3z"
        />
      </svg>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-link"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z"
        />
      </svg>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-image"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
        />
      </svg>
    </button>
    <button
      class="sc-bxivhb cshiNX ql-code-block"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.4 16.6L4.8 12l4.6-4.6L8 6l-6 6 6 6 1.4-1.4zm5.2 0l4.6-4.6-4.6-4.6L16 6l6 6-6 6-1.4-1.4z"
        />
      </svg>
    </button>
    <div
      class="mobile-placeholder"
    />
  </div>
</div>
`;
